src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp

```diff
@@ -57,15 +75,36 @@ IMergingAlgorithm::Status ReplacingSortedAlgorithm::merge()
     {
         SortCursor current = queue.current();
 
-        if (current->isLast() && skipLastRowFor(current->order))
+        if (sources[current.impl->order].chunk->empty() || (current->isLast() && skipLastRowFor(current->order)))
         {
+            auto & chunk = sources[current.impl->order].chunk;
+            if (!chunk->empty() && use_skipping_final)
+            {
+                if (selected_row.owned_chunk.get() == chunk.get())
+                {
+                    /// selected_row points to current source chunk but the chunk will be destroy soon, either in emitChunk() or queue.removeTop()
+                    /// In the first case, we create a cloned chunk with only one row from `selected_row.row_num` and let selected_row point to it
+                    /// In the second case, we mark selected_row.owned_chunk = nullptr
+                    /// In either case, the chunk is "lonely" and if later selected_row is inserted to final result, the chunk will be emitted
+                    /// immediately. This will create some blocks with only single row, but it's not a big problem.
```

Comment:
Honestly I don't want to squash chunks because the whole point of skipping algorithm is to prevent copy row-by-row.
But squashing into 1MB blocks doesn't look too bad. I will try and see the result.